### PR TITLE
Event in UIChooseClass to validate GTS class training

### LINKED ISSUE #814: Allow mods to modify the list of classes that can be trained in the GTS
Planning to add an event in UIChooseClass:GetClasses() where any mod can return a modified list. This allows GTS training to be restricted to certain classes without overriding the UI class.

### PR DESCRIPTION
**Pull request by ChrisTheThinMint**
Resolves #814.

### CODE REVIEW

**Comment by ChrisTheThinMint** [comment]
>     1. It's better to keep the return/`out` variable at position `0`
Why does this matter? I honestly don't see a difference.

>     2. IMO the event should not only add to the check, but allow listeners to override it completely - eg. what if someone wants to make `SoldierClassTemplate.NumInForcedDeck == 0` GTS-trainable?
Tbh i'm not sure how to do this. Could you point me to an example or provide some code?

>     3. IMO it's better to pass the template to the listeners, not the name. This will allow to save some boilerplate required to fetch the template in listeners
Makes sense. I can change that, no problem.

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/UIChooseClass.uc [hunk 1]
```diff
@@ -119,7 +119,7 @@ simulated function array<X2SoldierClassTemplate> GetClasses()
 	{		
 		SoldierClassTemplate = X2SoldierClassTemplate(Template);
 
-		if(TriggerGTSClassValidationEvent(SoldierClassTemplate.DataName) && SoldierClassTemplate.NumInForcedDeck > 0 && !SoldierClassTemplate.bMultiplayerOnly)
+		if(TriggerGTSClassValidationEvent(SoldierClassTemplate) && SoldierClassTemplate.NumInForcedDeck > 0 && !SoldierClassTemplate.bMultiplayerOnly)
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/UIChooseClass.uc]:
```suggestion
		if(TriggerGTSClassValidationEvent(SoldierClassTemplate))
```

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/UIChooseClass.uc [hunk 2]
```diff
@@ -139,29 +139,30 @@ simulated function array<X2SoldierClassTemplate> GetClasses()
 /// EventID: ValidateGTSClassTraining
 /// EventData: XComLWTuple {
 ///     Data: [
-///       in name SoldierClassName,
-///       out bool CanTrainClass
+///       out bool CanTrainClass,
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/UIChooseClass.uc]:
```suggestion
///       inout bool CanTrainClass,
```

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/UIChooseClass.uc [hunk 3]
```diff
@@ -139,29 +139,30 @@ simulated function array<X2SoldierClassTemplate> GetClasses()
...
+private function bool TriggerGTSClassValidationEvent(const X2SoldierClassTemplate SoldierClassTemplate)
 {
 	local XComLWTuple OverrideTuple;
 
 	OverrideTuple = new class'XComLWTuple';
 	OverrideTuple.Id = 'ValidateGTSClassTraining';
 	OverrideTuple.Data.Add(2);
+	// boolean to return
+	OverrideTuple.Data[0].kind = XComLWTVBool;
+	OverrideTuple.Data[0].b = true;
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/UIChooseClass.uc]:
```suggestion
	OverrideTuple.Data[0].b = SoldierClassTemplate.NumInForcedDeck > 0 && !SoldierClassTemplate.bMultiplayerOnly;
```

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/UIChooseClass.uc [hunk 4]
```diff
@@ -139,29 +139,30 @@ simulated function array<X2SoldierClassTemplate> GetClasses()
 /// EventID: ValidateGTSClassTraining
 /// EventData: XComLWTuple {
 ///     Data: [
-///       in name SoldierClassName,
-///       out bool CanTrainClass
+///       out bool CanTrainClass,
+///       in name SoldierClassName
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/UIChooseClass.uc]:
```suggestion
///       in X2SoldierClassTemplate SoldierClassTemplate
```

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/UIChooseClass.uc [hunk 5]
```diff
@@ -139,29 +139,30 @@ simulated function array<X2SoldierClassTemplate> GetClasses()
...
 ///     ]
 /// }
 /// EventSource: self (UIChooseClass)
 /// NewGameState: no
 /// ```
-private function bool TriggerGTSClassValidationEvent(const name SoldierClassName)
+private function bool TriggerGTSClassValidationEvent(const X2SoldierClassTemplate SoldierClassTemplate)
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/UIChooseClass.uc]: Why the `const` here?

**Comment by Xymanek** [comment]
> Could you point me to an example or provide some code?

See github suggestions above

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/UIChooseClass.uc (hunk 5 above)
Comment (ChrisTheThinMint) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/UIChooseClass.uc]: Not sure. I copied the event from one of paledbrook's commits. It doesnt seem to cause any issues, atleast.

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/UIChooseClass.uc [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+simulated function array<X2SoldierClassTemplate> GetClasses()
+{
+	local X2SoldierClassTemplateManager SoldierClassTemplateMan;
+	local X2SoldierClassTemplate SoldierClassTemplate;
+	local X2DataTemplate Template;
+	local array<X2SoldierClassTemplate> ClassTemplates;
+
+	SoldierClassTemplateMan = class'X2SoldierClassTemplateManager'.static.GetSoldierClassTemplateManager();
+
+	// Start Issue #814
```
Comment (Xymanek) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/UIChooseClass.uc]: This isn't the correct marking, e.g. you didn't change the loop.

You should move `// Start Issue #814` right before `/// HL-Docs` and add `// Issue #814` at the end of `if` line

**Comment by Xymanek** [comment]
Better compare: https://github.com/X2CommunityCore/X2WOTCCommunityHighlander/compare/d4eed8448094b6583201eda0ea73dea55ab9de3a...long-war-2:validate-gts-class-training

**Pull request merged**